codex-rs/core/src/codex.rs

```diff
@@ -469,6 +470,54 @@ impl Session {
         }
     }
 
+    async fn run_exec_with_events<'a>(
```

Comment:
This function returns `Result`, so can you add a docstring to comment that we guarantee that both `on_exec_command_begin()` and `on_exec_command_end()` are called, even if this function returns `Err`?

Also update to note that this is used both when the shell call is done originally as well as for the sandbox retry.